### ISSUE TITLE
Adaptations step influence on model graph 

### ISSUE THREAD

**Issue by JeremyAlain** [learnables/learn2learn#154]
Hi there, 
I have another quick question that is mostly related to pytorch itself. In the file `/learn2learn/algorithms/maml.py /`, specifically in the function `adapt` we call the following: 
```
grad_params = grad(loss,
                               diff_params,
                               retain_graph=second_order,
                               create_graph=second_order,
                               allow_unused=allow_unused)
```
As far as I understand it, `second_order=True` will keep the current graph after calculating the gradients. The use of this is to allow further gradient calculation through that graph. Put more simply, this is the important argument to enable second-order derivatives for MAML.

What I am wondering now is the following. If I have K-adaptation steps, as for example in the function `fast_adapt`of the file `learn2learn/examples/vision/maml_omniglot.py /`. I will obviously calculate k adaptation gradients and thus always call `grad` with `second_order=True`. Does this mean that each time the graph will be saved for further differentiation through it? And does this basically mean that if I do k adaptation steps, eventually when backpropagating for the meta-objective, I will backprobagate through k different adaptation steps? This basically means that I calculate a k'th order derivative right? 

Thanks

**Comment by seba-1511** [comment]
Hello again @JeremyAlain,

First, thanks for your complete answer on your previous thread, it helps clarify the issue.

The answer to your question is yes: PyTorch will keep the computation graphs and extend them as needed with the additional adaptation steps. When calling `.backward()`, the evaluation error will be back-propagated through each step.

However, you won't need higher-order derivatives -- only first and second. Intuitively you can see this as we never differentiate Hessians, only gradients. Note that PyTorch doesn't have to store the Hessians either, as the autograd library can make use of [efficient Hessian-vector products](https://justindomke.wordpress.com/2009/01/17/hessian-vector-products/).

Hope this helps.

**Comment by JeremyAlain** [comment]
Thanks again for the nice answer :).
